flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/ZooKeeperCompletedCheckpointStore.java

```diff
@@ -0,0 +1,310 @@
+	 */
+	public ZooKeeperCompletedCheckpointStore(
+			int maxNumberOfCheckpointsToRetain,
+			ClassLoader userClassLoader,
+			CuratorFramework client,
+			String checkpointsPath,
+			StateHandleProvider<CompletedCheckpoint> stateHandleProvider) throws Exception {
+
+		checkArgument(maxNumberOfCheckpointsToRetain >= 1, "Must retain at least one checkpoint.");
+
+		this.maxNumberOfCheckpointsToRetain = maxNumberOfCheckpointsToRetain;
+		this.userClassLoader = checkNotNull(userClassLoader, "User class loader");
+
+		checkNotNull(client, "Curator client");
+		checkNotNull(checkpointsPath, "Checkpoints path");
+		checkNotNull(stateHandleProvider, "State handle provider");
+
+		// Ensure that the checkpoints path exists
+		client.newNamespaceAwareEnsurePath(checkpointsPath)
+				.ensure(client.getZookeeperClient());
+
+		// All operations will have the path as root
+		this.client = client.usingNamespace(client.getNamespace() + checkpointsPath);
+
+		this.checkpointsInZooKeeper = new ZooKeeperStateHandleStore<>(
+				this.client, stateHandleProvider);
+
+		this.checkpointStateHandles = new ArrayDeque<>(maxNumberOfCheckpointsToRetain + 1);
+
+		LOG.info("Initialized in '{}'.", checkpointsPath);
+	}
+
+	/**
+	 * Gets the latest checkpoint from ZooKeeper and removes all others.
+	 *
+	 * <p><strong>Important</strong>: Even if there are more than one checkpoint in ZooKeeper,
+	 * this will only recover the latest and discard the others. Otherwise, there is no guarantee
+	 * that the history of checkpoints is consistent.
+	 */
+	@Override
+	public void recover() throws Exception {
+		LOG.info("Recovering checkpoints from ZooKeeper.");
+
+		// Get all there is first
+		List<ZooKeeperStateHandle<CompletedCheckpoint>> initialCheckpoints;
+		while (true) {
+			try {
+				initialCheckpoints = checkpointsInZooKeeper.getAllSortedByName();
+				break;
+			}
+			catch (ConcurrentModificationException e) {
+				LOG.warn("Concurrent modification while reading from ZooKeeper. Retrying.");
+			}
+		}
+
+		int numberOfInitialCheckpoints = initialCheckpoints.size();
+
+		LOG.info("Found {} checkpoints in ZooKeeper.", numberOfInitialCheckpoints);
+
+		if (numberOfInitialCheckpoints > 0) {
+			// Take the last one. This is the latest checkpoints, because path names are strictly
+			// increasing (checkpoint ID).
+			ZooKeeperStateHandle<CompletedCheckpoint> latest = initialCheckpoints
+					.get(numberOfInitialCheckpoints - 1);
+
+			CompletedCheckpoint latestCheckpoint = latest.getState(userClassLoader);
+
+			checkpointStateHandles.add(latest);
+
+			LOG.info("Initialized with {}. Removing all older checkpoints.", latestCheckpoint);
+
+			for (int i = 0; i < numberOfInitialCheckpoints - 1; i++) {
+				try {
+					removeFromZooKeeperAndDiscardCheckpoint(initialCheckpoints.get(i));
+				}
+				catch (Exception e) {
+					LOG.error("Failed to discard checkpoint", e);
+				}
+			}
+		}
+	}
+
+	@Override
+	public int getNextCheckpointID() throws Exception {
```

Comment:
If you don't find any checkpoints in the ZooKeeper, then simply start with `0`. Otherwise take the retrieved checkpoint ID from the latest completed checkpoint. Would this solve the problem?